File: solutions/cpp/Circular_queue.cpp
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <stdexcept>
#include <vector>

using std::cout;
using std::endl;
using std::exception;
using std::length_error;
using std::vector;

// @include
class Queue {
 public:
  explicit Queue(size_t cap) : data_(cap) {}

  void Enqueue(int x) {
    // Dynamically resizes due to data_.size() limit.
    if (size_ == data_.size()) {
      // Rearranges elements.
      rotate(data_.begin(), data_.begin() + head_, data_.end());
      head_ = 0, tail_ = size_;  // Resets head and tail.
      data_.resize(data_.size() * 2);
    }
    // Performs enqueue.
    data_[tail_] = x;
    tail_ = (tail_ + 1) % data_.size(), ++size_;
  }

  int Dequeue() {
    if (!size_) {
      throw length_error("empty queue");
    }
    --size_;
    int ret = data_[head_];
    head_ = (head_ + 1) % data_.size();
    return ret;
  }

  size_t size() const { return size_; }

 private:
  size_t head_ = 0, tail_ = 0, size_ = 0;
  vector<int> data_;
};
// @exclude

void Test() {
  Queue q(8);
  q.Enqueue(1);
  q.Enqueue(2);
  q.Enqueue(3);
  q.Enqueue(4);
  q.Enqueue(5);
  q.Enqueue(6);
  q.Enqueue(7);
  q.Enqueue(8);
  // Now head = 0 and tail = 0

  assert(1 == q.Dequeue());
  assert(2 == q.Dequeue());
  assert(3 == q.Dequeue());
  // Now head = 3 and tail = 0

  q.Enqueue(11);
  q.Enqueue(12);
  q.Enqueue(13);
  // Ok till here. Now head = 3 and tail = 3

  q.Enqueue(14);  // now the vector (data) is resized; but the head and tail.
                  // (or elements) does not change accordingly.
  q.Enqueue(15);
  q.Enqueue(16);
  q.Enqueue(17);
  q.Enqueue(18);
  // The elements starting from head=3 are overwritten!

  assert(4 == q.Dequeue());
  assert(5 == q.Dequeue());
  assert(6 == q.Dequeue());
  assert(7 == q.Dequeue());
  assert(8 == q.Dequeue());
  assert(11 == q.Dequeue());
  assert(12 == q.Dequeue());
}

int main(int argc, char* argv[]) {
  Test();
  Queue q(8);
  q.Enqueue(1);
  q.Enqueue(2);
  q.Enqueue(3);
  assert(1 == q.Dequeue());
  q.Enqueue(4);
  assert(2 == q.Dequeue());
  assert(3 == q.Dequeue());
  assert(4 == q.Dequeue());
  try {
    q.Dequeue();
  }
  catch (const exception& e) {
    cout << e.what() << endl;
  }
  // test resize().
  q.Enqueue(4);
  q.Enqueue(4);
  q.Enqueue(4);
  q.Enqueue(4);
  q.Enqueue(4);
  q.Enqueue(4);
  q.Enqueue(4);
  q.Enqueue(4);
  q.Enqueue(4);
  assert(q.size() == 9);
  return 0;
}

File: solutions/python/circular_queue.py
# @include
class Queue:
    SCALE_FACTOR = 2

    def __init__(self, capacity):
        self._entries = [None] * capacity
        self._head = self._tail = self._num_queue_elements = 0

    def enqueue(self, x):
        if self._num_queue_elements == len(self._entries):  # Needs to resize.
            # Makes the queue elements appear consecutively.
            self._entries = (
                self._entries[self._head:] + self._entries[:self._head])
            # Resets head and tail.
            self._head, self._tail = 0, self._num_queue_elements
            self._entries += [None] * (len(self._entries) * Queue.SCALE_FACTOR)

        self._entries[self._tail] = x
        self._tail = (self._tail + 1) % len(self._entries)
        self._num_queue_elements += 1

    def dequeue(self):
        if not self._num_queue_elements:
            raise IndexError('empty queue')
        self._num_queue_elements -= 1
        ret = self._entries[self._head]
        self._head = (self._head + 1) % len(self._entries)
        return ret

    def size(self):
        return self._num_queue_elements
# @exclude


def test():
    q = Queue(8)
    q.enqueue(1)
    q.enqueue(2)
    q.enqueue(3)
    q.enqueue(4)
    q.enqueue(5)
    q.enqueue(6)
    q.enqueue(7)
    q.enqueue(8)
    # Now head = 0 tail = 0

    assert 1 == q.dequeue()
    assert 2 == q.dequeue()
    assert 3 == q.dequeue()
    # head = 3 and tail = 0

    q.enqueue(11)
    q.enqueue(12)
    q.enqueue(13)
    # Ok till here. Now head = 3 and tail = 3

    q.enqueue(14)  # now the vector (data) is resized; but the head and tail.
    # (or elements) does not change accordingly.
    q.enqueue(15)
    q.enqueue(16)
    q.enqueue(17)
    q.enqueue(18)
    # The elements starting from head=3 are overwritten!

    assert 4 == q.dequeue()
    assert 5 == q.dequeue()
    assert 6 == q.dequeue()
    assert 7 == q.dequeue()
    assert 8 == q.dequeue()
    assert 11 == q.dequeue()
    assert 12 == q.dequeue()


def main():
    test()
    q = Queue(8)
    q.enqueue(1)
    q.enqueue(2)
    q.enqueue(3)
    assert 1 == q.dequeue()
    q.enqueue(4)
    assert 2 == q.dequeue()
    assert 3 == q.dequeue()
    assert 4 == q.dequeue()
    try:
        q.dequeue()
    except IndexError as e:
        print(e)
    # test resize.
    q.enqueue(4)
    q.enqueue(4)
    q.enqueue(4)
    q.enqueue(4)
    q.enqueue(4)
    q.enqueue(4)
    q.enqueue(4)
    q.enqueue(4)
    q.enqueue(4)
    assert q.size() == 9


if __name__ == '__main__':
    main()
